### PR TITLE
Bluespace Experi-Scanner

### PR DESCRIPTION
**Pull request by codehonky**
## About The Pull Request

Adds a bluespace experi-scanner to the game which unlocks with the Applied Bluespace tech node and can perform experiments instantly and from range.

## Why It's Good For The Game

The first thirty minutes of science gameplay in a shift is typically the exact same grindfest, and any roleplay from asking departments to open for early upgrades/experiment scans loses all glamour after the first few times a scientist performs it. The alternative to walking around with a bluespace scanner is just building copies of the machines for experiments in science anyway, which still lacks roleplay. QoL change for science that encourages walking outside the department more without counterhopping and allows it to engage in roleplay shticks sooner.

## Proof Of Testing

I ran and tested it locally.

## Changelog

:cl:
add: Added new bluespace experi-scanner.
image: Added a simple reskin of the experi-scanner for the bluespace variant.
/:cl:

### CODE REVIEW

**Comment by R3dtail** [comment]
Maybe consider making it a piece of RD's equipment?  I fully understand that people just do the meta strats nowadays with experiments (if they bother to do them at all), but you can make plenty of fun of showing up to people's departments to pester them about  being let in for experiments.  It's as fun as you make it.    Ranged experiment scanning has some niche uses that would be neat, but I also don't really like the idea of REDUCING the need to interact even if said interactions ARE typical and basically a script.  I think that's part of the charm and adds some believability to the whole job RP portion.  Just something to consider.

**Comment by aKromatopzia** [comment]
taking out the hypothetical situation that a scientist uses their experiments to trespass is both unfortunate, and amusing

**Comment by codehonky** [comment]
It barely takes away from the need to interact. Instead of waiting for doors to be opened, you still walk to where the things you'll be scanning are and can interact all the same but aren't forced to sit and wait for every door to open. It actually encourages walking and interacting more than building copies of the machines in science. And you can still use experiments as an excuse to trespass; just say there weren't materials for another bluespace scanner or you forgot to make one.

**Comment by R3dtail** [comment]
Y'know that's true.  I was approaching it from the aspect of the 'using through cameras" point brought up in the PR.  I don't mind the range part  since as you said, doesn't take away from interactions, just makes things slightly less annoying to an extent.

**Review by nevimer — CHANGES_REQUESTED** [review]
this all will need to be made either modular, or comment blocked for non-modular edits

**Comment by codehonky** [comment]
> this all will need to be made either modular, or comment blocked for non-modular edits

Done

**Comment by codehonky** [comment]
@nevimer The modular defines file was in the wrong spot - fixed and tested locally.

**Comment by codehonky** [comment]
@nevimer bumping this

File: modular_zubbers/code/modules/experisci/experiment/handlers/experiment_handler.dm [hunk 1]
```diff
@@ -0,0 +1,12 @@
+/datum/component/experiment_handler/proc/bs_ignored_handheld_experiment_attempt(datum/source, atom/target, mob/user, proximity_flag, params)
+	SIGNAL_HANDLER
+	if ((selected_experiment == null && !(config_flags & EXPERIMENT_CONFIG_ALWAYS_ACTIVE)) || config_flags & EXPERIMENT_CONFIG_SILENT_FAIL)
```
Comment (Majkl-J) [inline modular_zubbers/code/modules/experisci/experiment/handlers/experiment_handler.dm:3]:
```suggestion
	if ((isnull(selected_experiment) && !(config_flags & EXPERIMENT_CONFIG_ALWAYS_ACTIVE)) || config_flags & EXPERIMENT_CONFIG_SILENT_FAIL)
```

File: modular_zubbers/code/modules/experisci/experiment/handlers/experiment_handler.dm [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+		return .
+	bs_try_run_handheld_experiment(source, target, user, params)
+	. |= COMPONENT_AFTERATTACK_PROCESSED_ITEM
+	return .
+
+/datum/component/experiment_handler/proc/bs_try_run_handheld_experiment(datum/source, atom/target, mob/user, params)
+	SIGNAL_HANDLER
```
Comment (Majkl-J) [inline modular_zubbers/code/modules/experisci/experiment/handlers/experiment_handler.dm:10]: No need to have this when the parent that calls it already has it

```suggestion

```

File: modular_zubbers/code/modules/experisci/handheld_scanner.dm [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+	var/static/list/handheld_signals = list(
+		COMSIG_ITEM_PRE_ATTACK = TYPE_PROC_REF(/datum/component/experiment_handler, try_run_handheld_experiment),
+		COMSIG_ITEM_AFTERATTACK = TYPE_PROC_REF(/datum/component/experiment_handler, bs_ignored_handheld_experiment_attempt),
+	)
+	AddComponent(/datum/component/experiment_handler, \
+		allowed_experiments = list(/datum/experiment/scanning, /datum/experiment/physical), \
+		disallowed_traits = EXPERIMENT_TRAIT_DESTRUCTIVE, \
+		experiment_signals = handheld_signals, \
+		config_flags = EXPERIMENT_CONFIG_IMMEDIATE_ACTION|EXPERIMENT_CONFIG_WORKS_FROM_RANGE, \
+	)
```
Comment (Majkl-J) [inline modular_zubbers/code/modules/experisci/handheld_scanner.dm:19]:
```suggestion
	AddComponent(/datum/component/experiment_handler, \
		allowed_experiments = list(/datum/experiment/scanning, /datum/experiment/physical), \
		disallowed_traits = EXPERIMENT_TRAIT_DESTRUCTIVE, \
		config_flags = EXPERIMENT_CONFIG_IMMEDIATE_ACTION|EXPERIMENT_CONFIG_WORKS_FROM_RANGE, \
		experiment_signals = handheld_signals, \
	)
```

**Review by Majkl-J — COMMENTED** [review]
Looks testmerge ready, just minor code clarity touches

**Comment by pixelkitty286** [comment]
if this gets added i'm adding it to research borgs as an upgrade

**Comment by R3dtail** [comment]
After experiencing it a couple of times yeah no this was the greatest thing to have ever period.  Little powercreepy I guess but  hot damn it's sleek as fuck  and more interesting to use.

**Review by Majkl-J — APPROVED** [review]
I like it, the code is solid, nothing was broken during testing. If it feels too powercreepy later down the line we can move it to a different node.

**Pull request merged**